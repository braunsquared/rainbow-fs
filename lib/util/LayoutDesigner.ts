import { GUID } from '../model/Sitecore';

export interface IDeviceDefinition {
  ID: GUID;
  Layout: string;
  Placeholders: IPlaceholderDefinition[];
  Renderings: IRenderingDefinition[];
}

export interface ILayoutDefinition {
  Devices: IDeviceDefinition[];
}

export interface IRenderingDefinition {
  UID: GUID;
  ItemID: GUID;
  Cachable: string;
  Conditions: string;
  Datasource: string;
  MultiVariateTest: string;
  PersonalizationTest: string;
  Parameters: string;
  Placeholder: string;
  VaryByData: string;
  VaryByDevice: string;
  VaryByLogin: string;
  VaryByParameters: string;
  VaryByQueryString: string;
  VaryByUser: string;
  ClearOnIndexUpdate: string;
}

export interface IPlaceholderDefinition {
  UID: GUID;
  Key: string;
  MetaDataItemID: GUID;
}

export class LayoutDesigner {

}
