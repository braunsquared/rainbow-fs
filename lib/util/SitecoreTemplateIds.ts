
export const Folder = 'a87a00b1-e6db-45ab-8b54-636fec3b5523';
export const MediaFolder = 'fe5dd826-48c6-436d-b87a-7c4210c7413b';

export const Image = 'f1828a2c-7e5d-4bbd-98ca-320474871548';
export const Jpeg = 'daf085e8-602e-43a6-8299-038ff171349f';
