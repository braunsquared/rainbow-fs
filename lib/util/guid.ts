import { GUID } from '../model/Sitecore';

export function toGUID(val: string): GUID {
  const match = /^\{?([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})\}?$/.exec(val);
  if (match) {
    return match[1].toLowerCase();
  }
}

export function compareGUIDs(a: GUID, b: GUID): boolean {
  if (!a || !b) {
    return false
  } else {
    return a.toLowerCase() === b.toLowerCase();
  }
}