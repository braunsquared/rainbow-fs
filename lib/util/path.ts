
export const sep = '/';

export function normalize(path: string): string {
  const norm = path.replace(/\/{2,}/g, '/').replace(/\/$/, '');
  return norm.startsWith('/') ? norm : `/${norm}`;
}

export function join(...elements: (string | undefined)[] ): string {
  return normalize(elements.filter(i => !!i).join(sep));
}

export function split(path: string): string[] {
  const tokens = path.split(sep);
  return tokens.length > 0 && tokens[0] === '' ? tokens.slice(1) : tokens;
}

export interface ParsedPath {
  name: string,
  base: string,
  path: string,
  tokens: string[],
};

export function parse(path: string): ParsedPath {
  const norm = normalize(path);
  const tokens = split(norm);
  return {
    name: tokens[tokens.length - 1],
    base: join(...tokens.slice(0, tokens.length -1)),
    path: norm,
    tokens,
  };
}