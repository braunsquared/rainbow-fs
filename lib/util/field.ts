import { GUID, IField } from '../model/Sitecore';
import { toGUID } from './guid';

export function treelistValue(field: IField | undefined, defaultValue: GUID[] = []): GUID[] {
  if (!field || !field.Value) {
    return defaultValue;
  }

  return field.Value.split(/\n|\r\n/).map(id => toGUID(id)).filter(i => i);
}

export function booleanValue(field: IField | undefined, defaultValue: boolean = false): boolean {
  if (!field || !field.Value) {
    return defaultValue;
  }

  return field.Value === '1';
}