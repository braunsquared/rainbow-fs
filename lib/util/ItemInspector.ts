import { IStore } from '../Store';
import { IItem, GUID } from '../model/Sitecore';
import { treelistValue, booleanValue } from './field';
import { compareGUIDs } from './guid';

export interface FieldDefinition {
  ID: GUID;
  Name: string;
  Type: string;
  Shared: boolean;
  Unversioned: boolean;
  Item?: IItem;
}

export interface FieldDefinitionResults {
  errors: Error[];
  fields: FieldDefinition[];
}

class ItemInspector {
  static SystemTemplateGUIDs = [
    '1930bbeb-7805-471a-a3be-4858ac7cf696', // Standard Template
    'a87a00b1-e6db-45ab-8b54-636fec3b5523', // Folder
    'b36ba9fd-0dc0-49c8-bea2-e55d70e6af28', // JSS Route
  ];

  private _store: IStore;
  private _item: IItem;

  constructor(store: IStore, item: IItem) {
    this._store = store;
    this._item = item;
  }

  get Item(): IItem { return this._item; }

  getTemplateItem(): IItem | undefined {
    const db = this._store.getDatabase(this._item.DB);
    if (!db) {
      throw new Error(`Item database not found: ${this._item.DB}`);
    }

    return db.getItem(this._item.Template);
  }

  getStandardValues(): IItem | undefined {
    let standardVals = this._store.glob(`${this._item.Path.Path}/__Standard Values`, this._item.DB);
    standardVals = standardVals.filter(item => item.Parent === this._item.ID);
    return standardVals[0];
  }

  inspectTemplate(): ItemInspector {
    const tmpl = this.getTemplateItem();
    if (!tmpl) {
      throw new Error(`Template item not found: ${this._item.Template}`);
    }

    return new ItemInspector(this._store, tmpl);
  }

  children(): IItem[] {
    const children = this._store.glob(`${this._item.Path.Path}/*`, this._item.DB);
    return children.filter(i => i.Parent === this._item.ID);
  }

  fieldDefinitions(): FieldDefinitionResults {
    const result: FieldDefinitionResults = {
      errors: [],
      fields: [],
    };

    const db = this._store.getDatabase(this._item.DB)!;

    const generateFieldDefinitions = (inspector: ItemInspector) => {
      const sections = inspector.children();
      sections.forEach(sectionItem => {
        const itemInspect = new ItemInspector(this._store, sectionItem);
        const fields = itemInspect.children();
        fields.forEach(fieldItem => {
          const typeField = fieldItem.field('Type');
          const sharedField = fieldItem.field('Shared');
          const unversionField = fieldItem.field('Unversioned');

          result.fields.push({
            ID: fieldItem.ID,
            Name: fieldItem.Path.Name,
            Type: typeField && typeField.Value ? typeField.Value : '',
            Shared: booleanValue(sharedField),
            Unversioned: booleanValue(unversionField),
            Item: fieldItem,
          });
        });
      });
    }

    // Get the Base Templates
    const baseField = this._item.field('__Base template');
    if(baseField && baseField.Value) {
      const templateIds = treelistValue(baseField);
      const baseInspectors = templateIds.map(id => {
        const item = db.getItem(id);
        if (!item) {
          // Only push errors for items that aren't the standard template id
          if (!id || (id && ItemInspector.SystemTemplateGUIDs.indexOf(id.toLowerCase()) === -1)) {
            result.errors.push(new Error(`Unable to find base template: ${id}`));
          }
          return;
        }

        return new ItemInspector(this._store, item);
      });
      baseInspectors.forEach(inspect => inspect && generateFieldDefinitions(inspect));
    }

    // Get the field definitions for the template item
    generateFieldDefinitions(this);

    return result;
  }
}

export { ItemInspector };
