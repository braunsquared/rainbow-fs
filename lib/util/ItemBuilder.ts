import { IStore } from '../Store';
import { IItem, GUID } from '../model/Sitecore';
import { ItemInspector, FieldDefinition } from './ItemInspector';
import { compareGUIDs } from './guid';
import { Path } from '../impl/sitecore/Path';
import aguid from 'aguid';
import * as luxon from 'luxon';

export interface Blob {
  Value: string;
  BlobID: string;
}

export class ItemBuilder {
  private _store: IStore;
  private _templateId: GUID;

  private _fields: Map<FieldDefinition, any>;
  private _blobs: Map<FieldDefinition, Blob>;
  private _knownDefs: FieldDefinition[];

  constructor(store: IStore, templateItem: IItem | string) {
    this._store = store;
    this._fields = new Map();
    this._blobs = new Map();

    if (typeof templateItem === 'string') {
      this._templateId = templateItem;
      this._knownDefs = [];
    } else {
      this._templateId = templateItem.ID;
      const inspector = new ItemInspector(store, templateItem);
      this._knownDefs = inspector.fieldDefinitions().fields;
    }
  }

  setFieldByName(name: string, value: any): this {
    const def = this._knownDefs.find(def => def.Name === name);
    if (!def) {
      throw new Error(`Unknown field: ${name}`);
    }
    return this.setField(def, value);
  }

  setFieldByGUID(fieldId: GUID, value: any): this {
    const def = this._knownDefs.find(def => compareGUIDs(def.ID, fieldId));
    if (!def) {
      throw new Error(`Unknown field: ${fieldId}`);
    }
    return this.setField(def, value);
  }

  // tslint:disable-next-line:max-line-length
  setFieldWithDetails(fieldId: GUID, name: string, type: string, shared: boolean, unversioned: boolean, value: any): this {
    return this.setField(
      {
        ID: fieldId,
        Name: name,
        Type: type,
        Shared: shared,
        Unversioned: unversioned,
      },
      value,
    );
  }

  setField(def: FieldDefinition, value: any): this {
    for (const key of this._fields.keys()) {
      if (compareGUIDs(key.ID, def.ID) && key !== def) {
        throw new Error(`Duplicate field key: ${key.ID}`);
      }
    }

    this._fields.set(def, value);
    return this;
  }

  setBlobByName(name: string, base64: string): this {
    const def = this._knownDefs.find(def => def.Name === name);
    if (!def) {
      throw new Error(`Unknown field: ${name}`);
    }
    return this.setBlob(def, base64);
  }

  setBlobByGUID(fieldId: GUID, base64: string): this {
    const def = this._knownDefs.find(def => compareGUIDs(def.ID, fieldId));
    if (!def) {
      throw new Error(`Unknown field: ${fieldId}`);
    }
    return this.setBlob(def, base64);
  }

  // tslint:disable-next-line:max-line-length
  setBlobWithDetails(fieldId: GUID, name: string, type: string, shared: boolean, unversioned: boolean, base64: string): this {
    return this.setBlob(
      {
        ID: fieldId,
        Name: name,
        Type: type,
        Shared: shared,
        Unversioned: unversioned,
      },
      base64,
    );
  }

  setBlob(def: FieldDefinition, base64: string): this {
    const blobId = aguid(base64);
    this._blobs.set(def, { Value: base64, BlobID: blobId });

    return this;
  }

  private _serializeField = (def: FieldDefinition): any => {
    return {
      ID: def.ID,
      Hint: def.Name,
      Value: this._fields.get(def),
    };
  }

  private _serializeBlob = (def: FieldDefinition): any => {
    const value = this._blobs.get(def);

    if (!value) return;

    return {
      ID: def.ID,
      Hint: def.Name,
      BlobID: value.BlobID,
      Value: value.Value,
    };
  }

  private _serializeFields = (predicate: (def: FieldDefinition) => boolean): any[] => {
    const fieldDefs = Array.from(this._fields.keys()).filter(predicate).map(this._serializeField);
    const blobDefs = Array.from(this._blobs.keys()).filter(predicate).map(this._serializeBlob);
    return [...fieldDefs, ...blobDefs];
  }

  buildItem(dbName: string, scPath: string, language?: string): IItem {
    // We build the item js and then let the store item factory do the magic
    const path = new Path(scPath);

    const itemJs: any = {
      Name: path.Name,
      Template: this._templateId,
      Path: path.Path,
      DB: dbName,
    };

    const sharedFields = this._serializeFields(def => def.Shared);
    if (sharedFields.length) {
      itemJs.SharedFields = sharedFields;
    }

    if (language) {
      const langJs: any = { Language: language };
      const unversionedFields = this._serializeFields(def => def.Unversioned);
      if (unversionedFields.length) {
        langJs.Fields = unversionedFields;
      }

      const versionJs: any = { Version: 1 };
      const versionedFields = this._serializeFields(def => !def.Shared && !def.Unversioned);
      if (versionedFields.length) {
        versionJs.Fields = versionedFields;
      }
      langJs.Versions = [versionJs];
      itemJs.Languages = [langJs];
    }

    return this._store.createItemFromObject(itemJs, {
      isNew: true,
      isDirty: true,
      isVirtual: false,
    });
  }

  static formatDate(date: Date): string {
    return luxon.DateTime.fromJSDate(date).toFormat('yyyyMMdd\'T\'HHmmss\'Z\'');
  }
}
