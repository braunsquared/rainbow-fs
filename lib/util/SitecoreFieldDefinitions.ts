import { FieldDefinition } from './ItemInspector';

export const Created: FieldDefinition = {
  ID: '25bed78c-4957-4165-998a-ca1b52f67497',
  Name: '__Created',
  Type: 'datetime',
  Shared: false,
  Unversioned: false,
};

export const CreatedBy: FieldDefinition = {
  ID: '5dd74568-4d4b-44c1-b513-0af5f4cda34f',
  Name: '__Created by',
  Type: 'Single-Line Text',
  Shared: false,
  Unversioned: false,
};

export const SortOrder: FieldDefinition = {
  ID: 'ba3f86a2-4a1c-4d78-b63d-91c2779c1b5e',
  Name: '__Sortorder',
  Type: 'Single-Line Text',
  Shared: true,
  Unversioned: false,
};

export const FinalRenderings: FieldDefinition = {
  ID: '04bf00db-f5fb-41f7-8ab7-22408372a981',
  Name: '__Final Renderings',
  Type: 'Layout',
  Shared: false,
  Unversioned: false,
};

// Image Template Field Definitions

export const Image = {
  Icon: {
    ID: '06d5295c-ed2f-4a54-9bf2-26228d113318',
    Name: '__Icon',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  Blob: {
    ID: '40e50ed9-ba07-4702-992e-a912738d32dc',
    Name: 'Blob',
    Type: 'blob',
    Shared: true,
    Unversioned: false,
  },
  Extension: {
    ID: 'c06867fe-9a43-4c7d-b739-48780492d06f',
    Name: 'Extension',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  MimeType: {
    ID: '6f47a0a5-9c94-4b48-abeb-42d38def6054',
    Name: 'Mime Type',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  Size: {
    ID: '6954b7c7-2487-423f-8600-436cb3b6dc0e',
    Name: 'Size',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  Width: {
    ID: '22eac599-f13b-4607-a89d-c091763a467d',
    Name: 'Width',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  Height: {
    ID: 'de2ca9e4-c117-4c8a-a139-1ff4b199d15a',
    Name: 'Height',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  Dimensions: {
    ID: 'cb09946f-3218-4823-87d2-d5007c199a96',
    Name: 'Dimensions',
    Type: 'Single-Line Text',
    Shared: true,
    Unversioned: false,
  },
  Alt: {
    ID: '65885c44-8fcd-4a7f-94f1-ee63703fe193',
    Name: 'Alt',
    Type: 'Single-Line Text',
    Shared: false,
    Unversioned: false,
  },
};
