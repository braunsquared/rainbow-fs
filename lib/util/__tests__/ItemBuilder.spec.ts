import mockFs from 'mock-fs';
import { loadMockFs } from '../../../test/util';
import { Store } from '../../Store';
import { Bucket } from '../../Bucket';
import { ItemBuilder } from '../ItemBuilder';

describe('ItemBuilder', () => {
  let store: Store;
  let sampleData: any;

  beforeAll(() => {
    sampleData = loadMockFs();
  });

  beforeEach(() => {
    mockFs(sampleData);
    store = new Store();
    store.addBucket(new Bucket('/data/Templates', '/sitecore/templates/Helix/Project', 'master'));
    return store.readAll();
  });

  afterEach(() => {
    mockFs.restore();
  });

  it('should build an item from an item template', () => {
    const item = store.getDatabase('master')!.getItem('261918c5-521a-422d-ae59-23f1d39ee0a3')!;
    const builder = new ItemBuilder(store, item);
    builder.setFieldByName('heading', 'This is a test');
    builder.setFieldByName('content', 'of some cool stuff');
    const builtItem = builder.buildItem('master', '/sitecore/templates/Helix/Project/TestBuilder', 'en');
    mockFs.restore();
    expect(builtItem).toMatchSnapshot();
  });
});