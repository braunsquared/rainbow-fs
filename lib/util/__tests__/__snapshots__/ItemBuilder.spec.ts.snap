// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemBuilder should build an item from an item template 1`] = `
Item {
  "DB": "master",
  "ID": "a7c88f77-f55b-4f06-8ea7-dd8c5aae6dd0",
  "Languages": Array [
    Language {
      "Fields": Array [],
      "Language": "en",
      "Versions": Array [
        Version {
          "Fields": Array [
            Field {
              "Hint": "heading",
              "ID": "5813fa39-94c0-45b4-95fc-f22236aeb9a6",
              "Type": undefined,
              "Value": "This is a test",
            },
            Field {
              "Hint": "content",
              "ID": "aad95a88-cecc-4f18-ad44-2fdd0f2777cc",
              "Type": undefined,
              "Value": "of some cool shit",
            },
          ],
          "Version": 1,
        },
      ],
    },
  ],
  "Meta": Object {
    "isDirty": false,
    "isNew": true,
    "isVirtual": false,
  },
  "Parent": "0725a327-cd61-14f8-8aed-8d74b9917c6c",
  "Path": Path {
    "Folder": "/sitecore/templates/Helix/Project",
    "Name": "TestBuilder",
    "Path": "/sitecore/templates/Helix/Project/TestBuilder",
    "Tokens": Array [
      "sitecore",
      "templates",
      "Helix",
      "Project",
      "TestBuilder",
    ],
  },
  "SharedFields": Array [],
  "Template": "261918c5-521a-422d-ae59-23f1d39ee0a3",
}
`;
